File: src/Core/InstallCommand.cs
using System;
using System.Collections.Generic;
using System.IO;
using Newtonsoft.Json;
using Mono.Options;

namespace NPackage.Core
{
    public class InstallCommand : CommandBase
    {
        private static readonly JsonSerializer serializer = new JsonSerializer();
        private readonly DownloadWorkflow workflow = new DownloadWorkflow();
        private readonly string libPath = FindLibDirectory();
        private readonly string archivePath;
        private readonly string archiveDirectory;
        private readonly Dictionary<string, Package> packages = new Dictionary<string, Package>(StringComparer.InvariantCultureIgnoreCase);
        private bool logNeedsIndent;
        private int steps;

        public InstallCommand()
        {
            RepositoryUri = new Uri("http://np.partario.com/packages.js");

            workflow.Log += OnWorkflowLog;
            archivePath = Path.Combine(libPath, ".dist");
            archiveDirectory = archivePath + Path.DirectorySeparatorChar;
        }

        protected override OptionSet CreateOptionSet()
        {
            return base
                .CreateOptionSet()
                .Add("r|repository", "URL of the packages.js file", (Uri v) => RepositoryUri = v);
        }

        private void Log(string format, params object[] args)
        {
            string message = string.Format(format, args);

            Console.WriteLine(logNeedsIndent
                ? "        " + message
                : message);

            logNeedsIndent = true;
        }

        private void OnWorkflowLog(object sender, LogEventArgs e)
        {
            Log(e.Message);
        }

        public static string FindLibDirectory()
        {
            string path = "lib";
            string root = Path.Combine(Path.GetPathRoot(Environment.CurrentDirectory), "lib");
            while (!Directory.Exists(path))
            {
                string fullPath = Path.GetFullPath(path);
                if (string.Equals(root, fullPath, StringComparison.InvariantCultureIgnoreCase))
                    throw new InvalidOperationException("Couldn't find lib directory.");

                path = Path.Combine("..", path);
            }

            return path;
        }

        private Package FindPackage(string name)
        {
            Package package;
            if (packages.TryGetValue(name, out package))
                return package;
            else
            {
                string message = string.Format("There is no package called {0}.", name);
                throw new ArgumentException(message, "name");
            }
        }

        private void RegisterPackage(Package package)
        {
            packages[package.Name + "-" + package.Version] = package;

            Package latestPackage;
            if (!packages.TryGetValue(package.Name, out latestPackage) ||
                string.Compare(package.Version, latestPackage.Version, StringComparison.InvariantCultureIgnoreCase) > 0)
            {
                packages[package.Name] = package;
            }
        }

        private void RegisterPackage(string filename)
        {
            Package package;

            using (TextReader textReader = new StreamReader(filename))
            using (JsonReader jsonReader = new JsonTextReader(textReader))
                package = serializer.Deserialize<Package>(jsonReader);

            RegisterPackage(package);
        }

        private void RegisterRepository(string filename)
        {
            Repository repository;

            using (TextReader textReader = new StreamReader(filename))
            using (JsonReader jsonReader = new JsonTextReader(textReader))
                repository = serializer.Deserialize<Repository>(jsonReader);

            foreach (Package package in repository.Packages)
                RegisterPackage(package);

            foreach (Uri uri in repository.RepositoryImports)
                workflow.Enqueue(uri, archiveDirectory, RegisterRepository);

            foreach (Uri uri in repository.PackageImports)
                workflow.Enqueue(uri, archiveDirectory, RegisterPackage);
        }

        private void DownloadPackageContents(Package package)
        {
            Uri siteUri = package.MasterSites.Count > 0
                ? new Uri(RepositoryUri, package.MasterSites[0])
                : RepositoryUri;

            string packagePath = Path.Combine(libPath, Path.Combine(package.Name, package.Version));
            Directory.CreateDirectory(packagePath);

            foreach (KeyValuePair<string, Library> pair in package.Libraries)
            {
                Uri downloadUri = new Uri(siteUri, pair.Value.Binary);
                string filename = Path.Combine(packagePath, pair.Key);
                if (string.IsNullOrEmpty(downloadUri.Fragment))
                    workflow.Enqueue(downloadUri, filename, delegate { LogSuccess(downloadUri, filename); });
                else
                {
                    UriBuilder archiveUriBuilder = new UriBuilder(downloadUri) { Fragment = String.Empty };
                    Uri archiveUri = archiveUriBuilder.Uri;
                    workflow.Enqueue(archiveUri, archiveDirectory, archiveFilename => UnpackArchive(archiveFilename, downloadUri, filename));
                }
            }
        }

        private static void LogSuccess(Uri uri, string filename)
        {
            Console.WriteLine("\r ***    Installed {0} to {1}", uri, filename);
        }

        private void UnpackArchive(string archiveFilename, Uri uri, string filename)
        {
            FileInfo archiveFileInfo = new FileInfo(archiveFilename);
            FileInfo fileInfo = new FileInfo(filename);

            if (!fileInfo.Exists || archiveFileInfo.LastWriteTime > fileInfo.LastWriteTime)
            {
                Log("Unpacking {0} to {1}", archiveFilename, filename);

                string entryName = uri.Fragment.TrimStart('#');
                DownloadWorkflow.ExtractFile(archiveFilename, entryName, filename);

                LogSuccess(uri, filename);
            }
        }

        private void RunWorkflow()
        {
            do
            {
                steps++;
                Console.Write("[ {0,3} ] ", steps);
                logNeedsIndent = false;
            } while (workflow.Step());
        }

        protected override int RunCore()
        {
            Directory.CreateDirectory(archivePath);
            workflow.Enqueue(RepositoryUri, archiveDirectory, RegisterRepository);
            RunWorkflow();

            foreach (string name in PackageNames)
            {
                Package package = FindPackage(name);
                DownloadPackageContents(package);
            }

            RunWorkflow();
            return 0;
        }

        public Uri RepositoryUri { get; set; }

        public IList<string> PackageNames
        {
            get { return ExtraArguments; }
        }
    }
}

File: src/Core/InstallCommand.fs
namespace NPackage.Core

open System
open System.Collections.Generic
open System.IO
open Newtonsoft.Json

type InstallCommand() =
    inherit CommandBase()
    let mutable repositoryUri = new Uri("http://np.partario.com/packages.js")
    let mutable preview = false
    let baseUri = new Uri(Environment.CurrentDirectory + Path.DirectorySeparatorChar.ToString())
    let serializer = new JsonSerializer()
    let success = printfn "\r ***    Installed %s"
    let log = printfn "    --> %s"

    override this.CreateOptionSet() =
        base.CreateOptionSet()
            .Add("p|preview", "Only show what packages would be installed; don't install them", function
                                                                                                | null -> preview <- false
                                                                                                | _ -> preview <- true)
            .Add("r|repository=", "URL of the packages.js file", fun (v : string) -> this.RepositoryUri <- new Uri(baseUri, v))

    override this.RunCore() =
        let libPath = 
            let root = Path.Combine(Path.GetPathRoot(Environment.CurrentDirectory), "lib")
            let rec findLibDirectory path =
                if Directory.Exists(path) then
                    path
                else
                    let fullPath = Path.GetFullPath(path)
                    if String.Equals(root, fullPath, StringComparison.InvariantCultureIgnoreCase) then
                        raise (new InvalidOperationException("Couldn't find lib directory."))
                    else
                        findLibDirectory (Path.Combine("..", path))

            findLibDirectory "lib"

        let archivePath = Path.Combine(libPath, ".dist")
        let archiveDirectory = archivePath + Path.DirectorySeparatorChar.ToString()

        let rec buildGraphByName packages name =
            match Map.tryFind name packages with
            | Some { Metadata = { Package = package } } -> buildGraph packages package
            | None -> raise (new InvalidOperationException("There is no package called " + name + "."))

        and buildGraph packages package =
            let dependentOrder = package.Requires
                                 |> List.ofSeq
                                 |> List.collect (buildGraphByName packages)
                                 |> List.filter (fun p -> p <> package)
            List.append dependentOrder [package]

        let rec installPackageToPath packages packagePath (package : Package) = 
            let downloadLibrary name (library : Library) = Download.workflow {
                let libraryFilename = Path.Combine(packagePath, name)
                let uri = new Uri(new Uri(repositoryUri, package.MasterSites.[0]), library.Binary)

                let builder = new UriBuilder(uri)
                if builder.Fragment.Length > 0 then
                    let fragment = builder.Fragment.TrimStart('#')
                    builder.Fragment <- String.Empty
                    let! archiveFilename = Download.fetch builder.Uri archiveDirectory
                    let archiveFileInfo = new FileInfo(archiveFilename)
                    let libraryFileInfo = new FileInfo(libraryFilename)

                    if (not libraryFileInfo.Exists) || (archiveFileInfo.LastWriteTime > libraryFileInfo.LastWriteTime) then
                        log ("Unpacking " + archiveFilename + "#" + fragment)
                        Archive.extract archiveFilename fragment libraryFilename
                        success libraryFilename
                else
                    do! Download.fetch_ uri libraryFilename
                    success libraryFilename

                return 0
            }

            let copyLocal =
                if package.CopyLocal then
                    buildGraph packages package
                    |> List.filter (fun p -> p <> package)
                    |> List.map (installPackageToPath packages packagePath)
                else
                    [Download.workflow { return 0 }]

            Download.workflow {
                let! codes = package.Libraries
                            |> List.ofSeq
                            |> List.map (fun pair -> downloadLibrary pair.Key pair.Value)
                            |> List.append copyLocal
                            |> Download.batch

                return List.max codes
            }

        let installPackage packages (package : Package) =
            let packagePath = Path.Combine(Path.Combine(libPath, package.Name), package.Version)
            installPackageToPath packages packagePath package            

        let packages = Download.run { PackageGraph.download archiveDirectory repositoryUri with Log = log }
        let installOrder = this.PackageNames
                           |> List.ofSeq
                           |> List.collect (buildGraphByName packages)

        if preview then
            printfn "The following packages would be installed:\n"
            installOrder
            |> List.iter (fun p -> printfn "\t%s version %s" p.Name p.Version)
            0
        else
            { (installOrder
               |> List.map (installPackage packages)
               |> Download.batch) with Log = log }
            |> Download.run
            |> List.max

    member this.RepositoryUri
        with get() = repositoryUri
        and set(value) = repositoryUri <- value

    member this.PackageNames
        with get() = base.ExtraArguments

    member this.Preview
        with get() = preview
        and set(value) = preview <- value